crates/sui-core/src/notify_once.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use parking_lot::Mutex;
+use std::sync::Arc;
+use tokio::sync::futures::Notified;
+use tokio::sync::Notify;
+
+/// Notify once allows waiter to register for certain conditions and unblocks waiter
+/// when condition is signalled with `notify` method.
+///
+/// The functionality is somewhat similar to a tokio watch channel with subscribe method,
+/// however it is much less error prone to use NotifyOnce rather then tokio watch.
+///
+/// Specifically with tokio watch you may miss notification,
+/// if you subscribe to it after the value was changed
+/// (Note that this is not a bug in tokio watch, but rather a mis-use of it).
+///
+/// NotifyOnce guarantees that wait() will return once notify() is called,
+/// regardless of whether wait() was called before or after notify().
+pub struct NotifyOnce {
+    notify: Mutex<Option<Arc<Notify>>>,
+}
+
+impl NotifyOnce {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Notify all waiters, present and future about event
+    ///
+    /// After this method all pending and future calls to .wait() will return
+    ///
+    /// This method returns errors if called more then once
+    #[allow(clippy::result_unit_err)]
+    pub fn notify(&self) -> Result<(), ()> {
+        let Some(notify) = self.notify.lock().take()else { return Err(()) };
+        // At this point all `register` either registered with current notify,
+        // or will be returning immediately
+        notify.notify_waiters();
+        Ok(())
+    }
+
+    /// Awaits for `notify` method to be called.
+    ///
+    /// This future is cancellation safe.
+    pub async fn wait(&self) {
+        // Note that we only hold lock briefly when registering for notification
+        // There is a bit of a trickery here with lock - we take a lock and if it is not empty,
+        // we register .notified() first and then release lock
+        //
+        // This is to make sure no notification is lost because Notify::notify_waiters do not
+        // notify waiters that register **after** notify_waiters was called
+        let mut notify = None;
```

Comment:
Maybe call it lock_guard_opt?

Reply:
This variable holds instance of `Notify`, therefore the name choice